single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -468,8 +535,23 @@ def new_leader(self) -> None:
         if not self.state.get_keyfile():
             self.state.set_keyfile(self.workload.generate_keyfile())
 
-        # Sets the password for the system users
-        for user in (OperatorUser, BackupUser, MonitorUser, LogRotateUser):
+        if self.state.internal_user_passwords_are_initialized():
+            return
+
+        if system_users_secret_id := self.config.system_users:
+            user_passwords = None
+            try:
+                user_passwords = self.charm.state.get_secret_from_id(system_users_secret_id)
+                validate_charm_user_password_config(user_passwords)
+            except (ModelError, SecretNotFoundError, InvalidPasswordError) as e:
+                logger.error(f"Invalid system-users secret: {e}.")
+
+            if user_passwords:
+                for username, password in user_passwords.items():
+                    user = get_user_from_username(username)
+                    self.state.set_user_password(user, password)
+
```

Comment:
I would rework this in a less intricate and more natural way (which avoids back and forth):
```
user_passwords : dict[MongoDBUser, str] = {}
if system_users_secret_id := ...:
   // Same logic but store user, password in the dictionary

for user in Charm users:
    if not self.state.get_user_password(user):
		password = user_passwords.get(user, self.workload.generate_password())
        self.state.set_user_password(user, password)
```

This way:
 * You have less nested loops
 * You never iterate twice on the same user
 * You avoid having set password and then get password on the same user in the second loop
 * You never have to worry if `user_passwords is empty or not, they'll all be treated the same way once the map is built.